trin-core/src/portalnet/overlay.rs

```diff
@@ -277,11 +278,8 @@ where
                                     // Skip storing & propagating content if it's not valid
                                     .expect("Unable to validate received content: {err:?}");
 
-                                // Check if data should be stored, and store if true.
                                 // Ignore error since all validated content is propagated.
-                                let _ = storage
-                                    .write()
-                                    .store_if_should(&key, &content_value.to_vec());
+                                let _ = store.write().put(key.clone(), &content_value.to_vec());
```

Comment:
My concern with using `put()` here is that it's not obvious that the radius check is happening... which could lead to us performing a redundant radius check, or "putting" data into storage we expect to be stored.

imo, w/ databases generally "put" has a pretty cut & dry implication that we are simply "putting" data into the db, which isn't the case here.

Likewise, I believe the removed comment is still valid, and probably better off left in since (imo) it's not obvious that this check is happening.

Reply:
We can remove the radius check from `put`. Currently, [`store`](https://github.com/ethereum/trin/blob/8a9bb83a416aefbd0d095baa9346774d5ee97704/trin-core/src/portalnet/storage.rs#L164) behaves the same way. I would not be opposed to removing this check in `put`. If it falls outside of the radius, and the storage capacity is exceeded, then that data would be the farthest item and would be immediately purged. It might be useful to make the return type of `put` a list of evicted content keys.

As things are now, `put` may fail, and a possible error is `InsufficientRadius`, so a look at the API communicates that the radius may be a reason for failure. If that's not clear, and we want to keep the radius check there, then we can add more documentation.

For a key-value store, which is what the abstraction of the storage layer should be IMO, "get" and "put" are the canonical APIs (e.g. RocksDB).

Based on what we decide I can undelete the comment or add a new comment.

Reply:
It also doesn't seem like we're checking whether we already have the data here, which will get expensive when we're writing to disk.

I'm not totally against simplifying `put` to not do any checks, but I'm not sure about the idea of letting things outside of the radius be added. Especially in this scenario, when we're looping through a bunch of data and `put`ing it. Stored items outside of the radius will be immediately bumped, wasting the I/O necessary to have stored them because there was never an opportunity to later share them with the network.

> If it falls outside of the radius, and the storage capacity is exceeded

nit: The radius doesn't begin decreasing from `U256::MAX` until the storage capacity is reached, so there is no case where something's outside of the radius and storage capacity isn't exceeded.

Reply:
Yeah, without checks in `put`, it would be the responsibility of the caller to perform the check and avoid unnecessary I/O. I'm also curious on how we feel about `put` returning the list of evicted content keys/IDs @mrferris.

Reply:
@jacobkaufmann I meant that currently this code doesn't check whether an item is already in the DB. `PortalStorage::store` only checks radius. So currently it will write duplicate data, which is what I was referring to as expensive. I might be wrong, but I don't think that sqlite or rocksdb come with the optimization of not re-writing identical data. So I think that a `has_key` trait method (whose `PortalStorage` implementation uses the efficient `rocksdb::get_pinned`) is a good idea.

Which use-case do you have in mind for `put` returning evicted items?

Reply:
Right now I think it would mostly be useful for debugging. It would also be sort of similar to [`HashMap::insert`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.insert), which returns an `Option` that may hold the previous value for a particular key. That said, I don't think those are strong arguments in favor of the change. We can add it later if desired.